internal/jobcontainers/jobcontainer.go

```diff
@@ -447,3 +525,63 @@ func (c *JobContainer) IsOCI() bool {
 func (c *JobContainer) OS() string {
 	return "windows"
 }
+
+// For every process in the job `job`, run the function `work`. This can be used to grab/filter the SYSTEM_PROCESS_INFORMATION
+// data from every process in a job.
+func forEachProcessInfo(job *jobobject.JobObject, work func(*winapi.SYSTEM_PROCESS_INFORMATION)) error {
+	procInfos, err := systemProcessInformation()
+	if err != nil {
+		return err
+	}
+
+	pids, err := job.Pids()
+	if err != nil {
+		return err
+	}
+
+	pidsMap := make(map[uint32]struct{})
+	for _, pid := range pids {
+		pidsMap[pid] = struct{}{}
+	}
+
+	for _, procInfo := range procInfos {
+		if _, ok := pidsMap[uint32(procInfo.UniqueProcessID)]; ok {
+			work(procInfo)
+		}
+	}
+	return nil
+}
+
+// Get a slice of SYSTEM_PROCESS_INFORMATION for all of the processes running on the system.
+func systemProcessInformation() ([]*winapi.SYSTEM_PROCESS_INFORMATION, error) {
+	var (
+		systemProcInfo *winapi.SYSTEM_PROCESS_INFORMATION
+		procInfos      []*winapi.SYSTEM_PROCESS_INFORMATION
+		size           = uint32(1024 * 512)
+	)
+	for {
+		b := make([]byte, size)
+		systemProcInfo = (*winapi.SYSTEM_PROCESS_INFORMATION)(unsafe.Pointer(&b[0]))
+		status := winapi.NtQuerySystemInformation(
+			winapi.SystemProcessInformation,
+			uintptr(unsafe.Pointer(systemProcInfo)),
+			size,
+			&size,
+		)
+		if winapi.NTSuccess(status) {
+			break
+		} else if status != winapi.STATUS_INFO_LENGTH_MISMATCH {
+			return nil, winapi.RtlNtStatusToDosError(status)
+		}
+	}
+
+	for {
+		procInfos = append(procInfos, systemProcInfo)
+		if systemProcInfo.NextEntryOffset == 0 {
+			break
+		}
+		systemProcInfo = (*winapi.SYSTEM_PROCESS_INFORMATION)(unsafe.Pointer(uintptr(unsafe.Pointer(systemProcInfo)) + uintptr(systemProcInfo.NextEntryOffset)))
```

Comment:
It's unclear to me if this is a safe operation. According to the [unsafe package](https://golang.org/pkg/unsafe/) description:

```
Unlike in C, it is not valid to advance a pointer just beyond the end of its original allocation:

// INVALID: end points outside allocated space.
var s thing
end = unsafe.Pointer(uintptr(unsafe.Pointer(&s)) + unsafe.Sizeof(s))

// INVALID: end points outside allocated space.
b := make([]byte, n)
end = unsafe.Pointer(uintptr(unsafe.Pointer(&b[0])) + uintptr(n))
```

So I would think that since you previously cast `systemProcInfo` to a `SYSTEM_PROCESS_INFORMATION` pointer, then using pointer arithmetic with that beyond it's size would be invalid too even though we know that there's memory there that we've allocated and can access. IDK this may not be super relevant in this case.

Reply:
I had a couple iterations of this that failed checkptr, but this passes so I'm erring on the side of this is fine but I'm open to keep digging here

Reply:
My take is this is okay because the "original allocation" in this case is the `make([]byte, size)` on line 566, and we never advance past that allocation.

Reply:
On the other hand, it would probably be a good idea to validate each `NextEntryOffset` and make sure it's still within the bounds of `size`.

Reply:
@kevpar Done, sorry for delay